### PR TITLE
JSON encode() for storage

### PR DESCRIPTION
**Pull request by posgnu**

### CODE REVIEW

File: codechain/keystore/HDStorageJson.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    iv = os.urandom(16)
+
+    kdf = "pbkdf2"
+    kdf_params = {
+        "dklen": 32,
+        "salt": binascii.hexlify(salt).decode("ascii"),
+        "c": 262144,
+        "prf": "hmac-sha256",
+    }
+    derived_key = await pbkdf2_async(
```
Comment (foriequal0) [inline codechain/keystore/HDStorageJson.py]: Use the sync version

File: codechain/keystore/HDStorageJson.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+from ..crypto import get_public_from_private
+from .Errors import ErrorCode
+from .Errors import KeystoreError
+from .keys import key_from_public_key
+from .KeyType import KeyType
+from .Pbkdf2Async import pbkdf2_async
+from .Types import SecretStorage
+
+
+async def encode(seed: str, passphrase: str, meta: str):
```
Comment (foriequal0) [inline codechain/keystore/HDStorageJson.py]: Make it sync

File: codechain/keystore/HDStorageJson.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    kdf_params = {
+        "dklen": 32,
+        "salt": binascii.hexlify(salt).decode("ascii"),
+        "c": 262144,
+        "prf": "hmac-sha256",
+    }
+    derived_key = await pbkdf2_async(
+        passphrase, salt, kdf_params.get("c"), kdf_params.get("dklen"), "sha256"
+    )
+    cipher = AES.new(derived_key[:16], AES.MODE_EAX, iv=iv)
```
Comment (foriequal0) [inline codechain/keystore/HDStorageJson.py]: There also `AES.MODE_CTR`. Why did you use `AES.MODE_EAX`?

File: codechain/keystore/StorageJson.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+from ..crypto import get_public_from_private
+from .Errors import ErrorCode
+from .Errors import KeystoreError
+from .keys import key_from_public_key
+from .KeyType import KeyType
+from .Pbkdf2Async import pbkdf2_async
+from .Types import SecretStorage
+
+
+async def encode(private_key: str, key_type: KeyType, passphrase: str, meta: str):
```
Comment (foriequal0) [inline codechain/keystore/StorageJson.py]: Make it sync

File: codechain/keystore/StorageJson.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+    kdf = "pbkdf2"
+    kdf_params = {
+        "dklen": 32,
+        "salt": binascii.hexlify(salt).decode("ascii"),
+        "c": 262144,
+        "prf": "hmac-sha256",
+    }
+    derived_key = await pbkdf2_async(
+        passphrase, salt, kdf_params.get("c"), kdf_params.get("dklen"), "sha256"
```
Comment (foriequal0) [inline codechain/keystore/StorageJson.py]: Use the sync version instead of the async version.

File: codechain/keystore/StorageJson.py [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    )
+    cipher = AES.new(derived_key[:16], AES.MODE_EAX, iv=iv)
```
Comment (foriequal0) [inline codechain/keystore/StorageJson.py]: Check previous comment for `AES.MODE_CTR`

**Review by foriequal0 — CHANGES_REQUESTED** [review]

File: codechain/keystore/HDStorageJson.py (hunk 1 above)
Comment (posgnu) [reply codechain/keystore/HDStorageJson.py]: I changed it.

File: codechain/keystore/HDStorageJson.py (hunk 2 above)
Comment (posgnu) [reply codechain/keystore/HDStorageJson.py]: I changed it.

File: codechain/keystore/HDStorageJson.py (hunk 3 above)
Comment (posgnu) [reply codechain/keystore/HDStorageJson.py]: My mistake. Thx.

File: codechain/keystore/StorageJson.py (hunk 4 above)
Comment (posgnu) [reply codechain/keystore/StorageJson.py]: I changed it.

File: codechain/keystore/StorageJson.py (hunk 5 above)
Comment (posgnu) [reply codechain/keystore/StorageJson.py]: I changed it.

File: codechain/keystore/StorageJson.py (hunk 6 above)
Comment (posgnu) [reply codechain/keystore/StorageJson.py]: I changed it.

**Comment by posgnu** [comment]
@foriequal0 This is ready to be merged.

File: codechain/keystore/HDStorageJson.py [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+
+    kdf = "pbkdf2"
+    kdf_params = {
+        "dklen": 32,
+        "salt": binascii.hexlify(salt).decode("ascii"),
+        "c": 262144,
+        "prf": "hmac-sha256",
+    }
+    derived_key = pbkdf2(
+        passphrase, salt, kdf_params.get("c"), kdf_params.get("dklen"), "sha256"
```
Comment (foriequal0) [inline codechain/keystore/HDStorageJson.py]: Please prefer `[]` than `.get()`. `.get()` returns `None` silently without `KeyError` even if the dictionary is a local one, not an external one. Especially in dynamic typed language such as Python, it is prone to a bug by a typo.

**Review by foriequal0 — APPROVED** [review]

File: codechain/keystore/StorageJson.py [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+
+    kdf = "pbkdf2"
+    kdf_params = {
+        "dklen": 32,
+        "salt": binascii.hexlify(salt).decode("ascii"),
+        "c": 262144,
+        "prf": "hmac-sha256",
+    }
+    derived_key = pbkdf2(
+        passphrase, salt, kdf_params.get("c"), kdf_params.get("dklen"), "sha256"
```
Comment (foriequal0) [inline codechain/keystore/StorageJson.py]: Use `[]` rather than `.get()`

**Review by foriequal0 — CHANGES_REQUESTED** [review]

**Comment by foriequal0** [comment]
@GNUp Can you resolve the conflict? `pipenv install --dev -e .`? Rebase https://github.com/CodeChain-io/codechain-sdk-python/pull/32 upon this after resolving the conflict.
Also see this: https://github.com/CodeChain-io/codechain-sdk-python/pull/32/commits/db74fc76487de881c2ec5d06d8557a8e56af92d0#r341442171

**Comment by foriequal0** [comment]
Make the title more descriptive. You might add an object to the title.

File: codechain/keystore/HDStorageJson.py (hunk 7 above)
Comment (posgnu) [reply codechain/keystore/HDStorageJson.py]: Good point! I fixed them.

File: codechain/keystore/StorageJson.py (hunk 8 above)
Comment (posgnu) [reply codechain/keystore/StorageJson.py]: I changed them

**Comment by posgnu** [comment]
@foriequal0 Sorry, I don't understand what happens to lint action. I force-pushed this after running `pre-commit run --all-files` but the error still occurs.

**Comment by foriequal0** [comment]
This is changed by `pre-commit`
```diff
diff --git a/codechain/primitives/AssetAddress.py b/codechain/primitives/AssetAddress.py
index 1da7313..86e0fa7 100644
--- a/codechain/primitives/AssetAddress.py
+++ b/codechain/primitives/AssetAddress.py
@@ -1,9 +1,8 @@
 import re
+from dataclasses import dataclass
 from typing import List
 from typing import Union

-from dataclasses import dataclass
-
 from ..crypto import bech32_decode
 from ..crypto import bech32_encode
 from .HexString import H160
```

**Pull request merged**